File: NeetCode/Math_And_Geometry/Exponent/exponent.py
from typing import Optional, List


class Solution:
    def myPow(self, x: float, n: int) -> float:
        if x == 0:
            return 0
        if n == 0:
            return 1
        if n == 1:
            return x

        if n < 0:
            return 1 / self.myPow(x, abs(n))

        if n % 2 == 0:
            res = self.myPow(x, n // 2)
            return res * res

        return self.myPow(x, n - 1) * x


if __name__ == "__main__":
    sol = Solution()
    x = 2.00000
    n = 5
    print(sol.myPow(x, n))
    print("Running Solution...")

File: NeetCode/Math_And_Geometry/Exponent/Exponent.cpp
// Exponent.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    double myPow(double x, int n)
    {
        if (x == 0)
        {
            return 0;
        }
        if (n == 0)
        {
            return 1;
        }
        if (n == 1)
        {
            return x;
        }

        if (n < 0)
        {
            return 1.0 / myPow(x, abs(n));
        }

        if (n % 2 == 0)
        {
            double result = myPow(x, n / 2);
            return result * result;
        }

        return myPow(x, n - 1) * x;
    }
};
